htlcswitch/link.go

```diff
@@ -2384,8 +2521,22 @@ func (l *channelLink) handleUpstreamMsg(msg lnwire.Message) {
 			}
 		}
 
+		// If we can send updates then we can process adds in case we
+		// are the exit hop and need to send back resolutions, or in
+		// case there are validity issues with the packets. Otherwise
+		// we defer the action until resume.
+		//
+		// We are free to process the settles and fails without this
+		// check since processing those can't result in further updates
+		// to this channel link.
+		if l.quiescer.canSendUpdates() {
```

Comment:
because the exit hop processing might cause us to redeem an HTLC which would be a spec violation. This would be solved if we fixed the division of responsibility I detailed in #8634